libfwupdplugin/fu-dfuse-firmware-image.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (C) 2015 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+/**
+ * SECTION:fu-dfuse-firmware-image
+ * @short_description: Object representing a a firmware image
+ *
+ * A #FuDfuseFirmwareImage is typically made up of several #FuChunk objects,
+ * although there may only be one.
+ *
+ * See also: #FuChunk
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-chunk-private.h"
+#include "fu-common.h"
+
+#include "fu-dfuse-firmware-image.h"
+
+static void fu_dfuse_firmware_image_finalize			 (GObject *object);
+
+typedef struct {
+	GPtrArray		*chunks;
+	gchar			 name[255];
```

Comment:
this seems a bit wasteful.  can't you just create it as needed and check the bounds on set?

Reply:
I can do better than that; the only reason we were copying the 255 bytes of junk was because we were round-tripping that nonfree dfuse file. Now we can build our own, we don't have to blindly copy the array as we know it's all NUL. I folded the name into the FuFirmwareImage::id and just moved the chunking there too -- removing half a dozen of the new symbols in the process! Review of that new code very welcome.